cpp/serve/request_state.h

```diff
@@ -0,0 +1,113 @@
+/*!
+ *  Copyright (c) 2023 by Contributors
+ * \file serve/request_state.h
+ * \brief The data structure maintaining the generation states of user requests.
+ */
+#ifndef MLC_LLM_SERVE_REQUEST_STATE_H_
+#define MLC_LLM_SERVE_REQUEST_STATE_H_
+
+#include <tvm/runtime/container/array.h>
+#include <tvm/runtime/ndarray.h>
+#include <tvm/runtime/object.h>
+
+#include "config.h"
+#include "data.h"
+
+namespace mlc {
+namespace llm {
+namespace serve {
+
+using namespace tvm::runtime;
+
+/*!
+ * \brief The state of a request with regard to some single model.
+ * \details In MLC LLM, the serving engine may leverage multiple models
+ * to fulfill a user generation request (e.g., use speculation decoding).
+ * For each request, we isolate its states (e.g. the generated tokens)
+ * on each model. This is to say, we use RequestModelState to store
+ * the state of a user request on a single model (rather than all models).
+ */
+class RequestModelStateNode : public Object {
+ public:
+  /*!
+   * \brief The corresponding request id of this state.
+   * It is the **physical index** of the request in the running request queue.
+   * If the request is on hold (not in the running queue), the request id
+   * should be -1.
+   */
+  int request_id = -1;
+  /*! \brief The corresponding model id of this state. */
+  int model_id = -1;
+  /*!
+   * \brief The committed generated token ids. A token is "committed"
+   * means it will no longer be updated (or changed).
+   */
+  std::vector<int32_t> committed_tokens;
```

Comment:
I guess beam search is not the primary concern at the moment, but I wonder how this would work with beam search if related.

Reply:
beam search would needs an update of the state containing parallel streams.